corehq/apps/users/forms.py

```diff
@@ -1273,17 +1274,26 @@ def clean_domains(self):
 
 class CreateDomainPermissionsMirrorForm(forms.Form):
     mirror_domain = forms.CharField(label=ugettext_lazy('Project Space'), max_length=30, required=True)
-
     def __init__(self, *args, **kwargs):
         if 'domain' not in kwargs:
             raise Exception('Expected kwargs: domain')
         self.domain = kwargs.pop('domain', None)
         super().__init__(*args, **kwargs)
 
     def clean_mirror_domain(self):
-        mirror_domain = self.data.get('mirror_domain')
-        if self.domain == mirror_domain:
+        mirror_domain_name = self.data.get('mirror_domain')
+        if self.domain == mirror_domain_name:
             raise forms.ValidationError(_("""
                 Enterprise permissions cannot be granted from a project space to itself.
             """))
-        return mirror_domain
+        mirror_domain = Domain.get_by_name(mirror_domain_name)
+        if mirror_domain is not None:
+            try:
+                mirror = DomainPermissionsMirror(source=self.domain, mirror=mirror_domain)
+                mirror.save()
+            except IntegrityError:
+                message = '"' + mirror_domain_name + '"' + ' has already been added.'
+                raise forms.ValidationError(_(message))
+        else:
+            raise forms.ValidationError(_('Please enter valid project space.'))
```

Comment:
sorry for being vague earlier - I do think the save action belongs elsewhere, probably in the main view or in a `save` method on the form (either is fine).  Expectation would be that `form.is_valid` only checks validity, and doesn't yet commit any changes.

You could do something like this here, then create the mirror in the view:
```suggestion
        mirror_domain = Domain.get_by_name(mirror_domain_name)
        if not mirror_domain:
            raise forms.ValidationError(_('Please enter valid project space.'))
        if DomainPermissionsMirror.objects.filter(mirror=mirror_domain).exists():
            message = _('"{mirror_domain_name}" has already been added.')
            raise forms.ValidationError(message.format(mirror_domain_name=mirror_domain_name))
```
(it looks like you accidentally reverted the translation change Jenny mentioned - added that back too)

If you're concerned about needing to duplicate the db constraint checks, a `ModelForm` would add that validation to `is_valid` automatically, though that'd take a bit more rejiggering of this view and form, so I probably wouldn't bother.

Reply:
No worries, thanks for bearing with me while I get familiar with the code. I made these changes.